File: python/compress_zip/attention.py
"""
Grouped Query Attention (GQA/MQA) with RoPE.

Matches CUDA attention.cu kernel.
"""

import numpy as np
from typing import Optional
from dataclasses import dataclass

from .primitives import sra_rne_tte_s32, sra_rne_tte_s64_to_s32, clamp_i8, div_rne_tte_s64_to_s32, sra_rne_tte_s32_np
from .lut import get_rope_lut, get_exp2_lut, ROPE_HALF_DIM

# Constants matching CUDA attention.cu
HEAD_DIM = 64
MAX_SEQ_LEN = 64
TOKENS_PER_BATCH = 16

# CUDA constants for attention score scaling
Q_SHIFT = 2
SQRT_HEAD_DIM_SHIFT = 3

# Score to exp mapping coefficient for Q0.14 scores (matches CUDA exactly)
# LOG2E = 1.4426950408889634
# ATTN_COEF = LOG2E / 64.0
# ATTN_COEF_Q24 = round(ATTN_COEF * 2^24)
LOG2E = 1.4426950408889634
ATTN_COEF = LOG2E / 64.0
ATTN_COEF_Q24 = int(ATTN_COEF * (1 << 24) + 0.5)  # 378173


def exp_q16_from_attn_diff(diff: int, exp_lut) -> int:
    """
    Compute exp2 for attention score difference.
    Matches CUDA exp_q16_from_attn_diff exactly.

    diff should be <= 0 (score - max_score)
    Returns uint16 Q16 representation.
    """
    if diff >= 0:
        return 65535

    neg = -diff

    # t256 = round(neg * ATTN_COEF_Q24 / 2^24)
    t256 = (neg * ATTN_COEF_Q24 + (1 << 23)) >> 24

    ip = t256 >> 8  # integer part
    frac = t256 & 255  # fractional part

    if ip >= 31:
        return 1

    base = (1 << 16) >> ip

    # Get fractional multiplier from LUT
    if hasattr(exp_lut, '__getitem__'):
        # numpy array or Exp2LutQ16 class
        frac_mul = int(exp_lut[frac])
    else:
        frac_mul = int(exp_lut[frac])

    out = (base * frac_mul + 0x8000) >> 16

    if out < 1:
        out = 1
    if out > 65535:
        out = 65535

    return out


@dataclass
class KVCache:
    """Key-Value cache for autoregressive generation."""
    k_cache: np.ndarray  # [max_seq_len, head_dim] as int8
    v_cache: np.ndarray  # [max_seq_len, head_dim] as int8
    seq_len: int  # Current sequence length

    @classmethod
    def create(cls, max_seq_len: int = MAX_SEQ_LEN, head_dim: int = HEAD_DIM) -> "KVCache":
        return cls(
            k_cache=np.zeros((max_seq_len, head_dim), dtype=np.int8),
            v_cache=np.zeros((max_seq_len, head_dim), dtype=np.int8),
            seq_len=0,
        )

    def append(self, k: np.ndarray, v: np.ndarray):
        """Append new K/V to cache."""
        if k.ndim == 1:
            k = k.reshape(1, -1)
            v = v.reshape(1, -1)

        n_new = k.shape[0]
        assert self.seq_len + n_new <= self.k_cache.shape[0], "Cache overflow"

        self.k_cache[self.seq_len:self.seq_len + n_new] = k
        self.v_cache[self.seq_len:self.seq_len + n_new] = v
        self.seq_len += n_new


def apply_rope_i8(
    x: np.ndarray,
    pos: int,
    rope_cos: np.ndarray = None,
    rope_sin: np.ndarray = None,
) -> np.ndarray:
    """
    Apply RoPE to a single vector at position pos.

    x is [head_dim] as int8.
    rope_cos/rope_sin: [max_seq_len, half_dim] int16 LUTs (optional, generated if not provided)
    Returns rotated vector as int8.
    """
    # Use embedded LUTs if provided, otherwise generate
    if rope_cos is None or rope_sin is None:
        rope_lut = get_rope_lut()
        use_class = True
    else:
        use_class = False

    out = np.zeros_like(x)

    for i in range(ROPE_HALF_DIM):
        # Get cos/sin from LUT (Q15)
        if use_class:
            cos_val = rope_lut.get_cos(pos, i)
            sin_val = rope_lut.get_sin(pos, i)
        else:
            cos_val = int(rope_cos[pos, i])
            sin_val = int(rope_sin[pos, i])

        # Get pair of values
        x0 = int(x[i])
        x1 = int(x[i + ROPE_HALF_DIM])

        # Rotation: [cos, -sin; sin, cos]
        # out[i] = x0 * cos - x1 * sin
        # out[i+half] = x0 * sin + x1 * cos

        # x is Q0.7, cos/sin is Q1.15
        # Product is Q1.22, shift by 15 to get Q0.7
        out0 = x0 * cos_val - x1 * sin_val
        out1 = x0 * sin_val + x1 * cos_val

        out[i] = clamp_i8(sra_rne_tte_s32(out0, 15))
        out[i + ROPE_HALF_DIM] = clamp_i8(sra_rne_tte_s32(out1, 15))

    return out.astype(np.int8)


def apply_rope_q_i8(
    q: np.ndarray,
    pos: int,
    rope_cos: np.ndarray = None,
    rope_sin: np.ndarray = None,
) -> np.ndarray:
    """
    Apply RoPE to Q with additional scaling (shift by Q_SHIFT).

    This prevents overflow in attention scores.
    rope_cos/rope_sin: [max_seq_len, half_dim] int16 LUTs (optional, generated if not provided)
    """
    # Use embedded LUTs if provided, otherwise generate
    if rope_cos is None or rope_sin is None:
        rope_lut = get_rope_lut()
        use_class = True
    else:
        use_class = False

    out = np.zeros_like(q)

    for i in range(ROPE_HALF_DIM):
        if use_class:
            cos_val = rope_lut.get_cos(pos, i)
            sin_val = rope_lut.get_sin(pos, i)
        else:
            cos_val = int(rope_cos[pos, i])
            sin_val = int(rope_sin[pos, i])

        x0 = int(q[i])
        x1 = int(q[i + ROPE_HALF_DIM])

        out0 = x0 * cos_val - x1 * sin_val
        out1 = x0 * sin_val + x1 * cos_val

        # CUDA: shift by 15 first, then by NET_SHIFT (two separate rounding operations)
        # NET_SHIFT = SQRT_HEAD_DIM_SHIFT - Q_SHIFT = 3 - 2 = 1
        x_rot = sra_rne_tte_s32(out0, 15)
        y_rot = sra_rne_tte_s32(out1, 15)

        # Apply NET_SHIFT = 1
        x_rot = sra_rne_tte_s32(x_rot, 1)
        y_rot = sra_rne_tte_s32(y_rot, 1)

        out[i] = clamp_i8(x_rot)
        out[i + ROPE_HALF_DIM] = clamp_i8(y_rot)

    return out.astype(np.int8)


def gqa_attention_mqa_i8(
    q: np.ndarray,
    k: np.ndarray,
    v: np.ndarray,
    pos: int,
    kv_cache: Optional[KVCache] = None,
    exp2_lut: np.ndarray = None,
    rope_cos: np.ndarray = None,
    rope_sin: np.ndarray = None,
    score_mul_q15: int = 32768,
) -> np.ndarray:
    """
    Multi-Query Attention (MQA) with single K/V head.
    Matches CUDA gqa_attention_mqa_i8_kernel exactly.

    Args:
        q: Query vector [head_dim] as int8
        k: Key vector [head_dim] as int8
        v: Value vector [head_dim] as int8
        pos: Current position (0-indexed)
        kv_cache: Optional KV cache for autoregressive generation
        exp2_lut: Optional [256] uint16 exp2 LUT (uses generated if not provided)
        rope_cos: Optional [max_seq_len, half_dim] int16 RoPE cos LUT
        rope_sin: Optional [max_seq_len, half_dim] int16 RoPE sin LUT
        score_mul_q15: Score multiplier in Q0.15 format (default 32768 = 1.0)

    Returns:
        Output vector [head_dim] as int8
    """
    # Use embedded LUTs if provided, otherwise generate
    if exp2_lut is None:
        exp_lut = get_exp2_lut()
    else:
        exp_lut = exp2_lut

    # Apply RoPE
    q_rot = apply_rope_q_i8(q, pos, rope_cos, rope_sin)
    k_rot = apply_rope_i8(k, pos, rope_cos, rope_sin)

    # Update KV cache if provided
    if kv_cache is not None:
        kv_cache.k_cache[pos] = k_rot
        kv_cache.v_cache[pos] = v
        seq_len = pos + 1
        k_cache = kv_cache.k_cache[:seq_len]
        v_cache = kv_cache.v_cache[:seq_len]
    else:
        # No cache, just current token
        seq_len = 1
        k_cache = k_rot.reshape(1, -1)
        v_cache = v.reshape(1, -1)

    # Compute attention scores matching CUDA exactly (vectorized)
    q_rot_64 = q_rot.astype(np.int64)
    k_cache_64 = k_cache.astype(np.int64)
    dots = np.sum(q_rot_64 * k_cache_64, axis=1)  # [seq_len]

    # CUDA: dot_unshift = round(dot >> Q_SHIFT) with ties-to-even
    dots_unshift = sra_rne_tte_s32_np(dots, Q_SHIFT)

    # CUDA: prod = dot_unshift * score_mul; score = round(prod >> 15) with ties-to-even
    prods = dots_unshift.astype(np.int64) * score_mul_q15
    scores = np.array([sra_rne_tte_s64_to_s32(int(p), 15) for p in prods])

    # Online softmax (matching CUDA)
    m = -(1 << 30)  # Running max (use large negative instead of INT32_MIN)
    s = 0  # Running sum of exp weights (uint64)
    vacc = np.zeros(HEAD_DIM, dtype=np.int64)  # Weighted V accumulator

    for t in range(seq_len):
        score = int(scores[t])

        # Update online softmax
        if score > m:
            # New max found - rescale
            if m != -(1 << 30):
                diff_old = m - score
                scale_old = exp_q16_from_attn_diff(diff_old, exp_lut)
                # Rescale S: (S * scale + 0x8000) >> 16
                s = (s * scale_old + 0x8000) >> 16
                # Rescale vacc (vectorized, matching CUDA exactly)
                prod = vacc * scale_old
                adjustment = np.where(prod >= 0, 1 << 15, -(1 << 15))
                vacc = (prod + adjustment) >> 16
            m = score

        # Compute exp weight for this token
        diff = score - m
        w = exp_q16_from_attn_diff(diff, exp_lut)
        s += w

        # Accumulate weighted V (vectorized)
        vacc += w * v_cache[t].astype(np.int64)

    # Normalize output
    denom = s if s != 0 else 1

    # Vectorized output computation
    y = np.zeros(HEAD_DIM, dtype=np.int64)
    for i in range(HEAD_DIM):
        y[i] = div_rne_tte_s64_to_s32(int(vacc[i]), denom)
    output = np.clip(y, -128, 127).astype(np.int8)

    return output


def gqa_attention_mqa_i8_cached(
    q: np.ndarray,
    k_cache: np.ndarray,
    v_cache: np.ndarray,
    pos: int,
    exp2_lut: np.ndarray = None,
    rope_cos: np.ndarray = None,
    rope_sin: np.ndarray = None,
    score_mul_q15: int = 32768,
    sink_key: np.ndarray = None,
) -> np.ndarray:
    """
    Multi-Query Attention with pre-cached K/V.
    Matches CUDA exactly - K is already rotated and stored in cache.

    Args:
        q: Query vector [head_dim] as int8
        k_cache: Key cache [seq_len, head_dim] as int8 (already RoPE'd)
        v_cache: Value cache [seq_len, head_dim] as int8
        pos: Current position (0-indexed)
        exp2_lut: Optional [256] uint16 exp2 LUT
        rope_cos: Optional [max_seq_len, half_dim] int16 RoPE cos LUT
        rope_sin: Optional [max_seq_len, half_dim] int16 RoPE sin LUT
        score_mul_q15: Score multiplier in Q0.15 format
        sink_key: Optional [head_dim] int8 attention sink key (value is zero)

    Returns:
        Output vector [head_dim] as int8
    """
    if exp2_lut is None:
        exp_lut = get_exp2_lut()
    else:
        exp_lut = exp2_lut

    # Apply RoPE to Q only (K is already rotated in cache)
    q_rot = apply_rope_q_i8(q, pos, rope_cos, rope_sin)

    seq_len = pos + 1

    # Compute attention scores matching CUDA exactly (vectorized)
    # Q*K dot products for all positions at once
    q_rot_64 = q_rot.astype(np.int64)
    k_cache_64 = k_cache[:seq_len].astype(np.int64)
    dots = np.sum(q_rot_64 * k_cache_64, axis=1)  # [seq_len]

    # CUDA: dot_unshift = round(dot >> Q_SHIFT) with ties-to-even
    dots_unshift = sra_rne_tte_s32_np(dots, Q_SHIFT)

    # CUDA: prod = dot_unshift * score_mul; score = round(prod >> 15) with ties-to-even
    prods = dots_unshift.astype(np.int64) * score_mul_q15
    # For 64-bit values, process element-wise to ensure correct rounding
    scores = np.array([sra_rne_tte_s64_to_s32(int(p), 15) for p in prods])

    # Compute sink score if sink key is provided
    sink_score = None
    if sink_key is not None:
        # Dot product with RoPE'd Q (no RoPE on sink key)
        sink_dot = int(np.sum(q_rot.astype(np.int64) * sink_key.astype(np.int64)))
        sink_dot_unshift = sra_rne_tte_s32_np(np.array([sink_dot]), Q_SHIFT)[0]
        sink_prod = int(sink_dot_unshift) * score_mul_q15
        sink_score = sra_rne_tte_s64_to_s32(sink_prod, 15)

    # Online softmax (matching CUDA)
    m = -(1 << 30)
    s = 0
    vacc = np.zeros(HEAD_DIM, dtype=np.int64)

    # Include sink score in initial max if present
    if sink_score is not None:
        m = sink_score
        w_sink = exp_q16_from_attn_diff(0, exp_lut)  # diff = 0 since score == m
        s = w_sink
        # Note: sink value is zero, so no contribution to vacc

    for t in range(seq_len):
        score = int(scores[t])

        if score > m:
            if m != -(1 << 30):
                diff_old = m - score
                scale_old = exp_q16_from_attn_diff(diff_old, exp_lut)
                s = (s * scale_old + 0x8000) >> 16
                # Vectorized vacc rescaling matching CUDA
                prod = vacc * scale_old
                adjustment = np.where(prod >= 0, 1 << 15, -(1 << 15))
                vacc = (prod + adjustment) >> 16
            m = score

        diff = score - m
        w = exp_q16_from_attn_diff(diff, exp_lut)
        s += w

        vacc += w * v_cache[t].astype(np.int64)

    denom = s if s != 0 else 1

    # Vectorized output computation
    vacc_arr = vacc.astype(np.int64)
    y = np.zeros(HEAD_DIM, dtype=np.int64)
    for i in range(HEAD_DIM):
        y[i] = div_rne_tte_s64_to_s32(int(vacc_arr[i]), denom)
    output = np.clip(y, -128, 127).astype(np.int8)

    return output


# Test cases
if __name__ == "__main__":
    # Test RoPE
    x = np.zeros(HEAD_DIM, dtype=np.int8)
    x[0] = 64  # First element non-zero

    # At position 0, rotation should be minimal
    x_rot = apply_rope_i8(x, 0)
    print(f"RoPE at pos 0: {x[:4]} -> {x_rot[:4]}")

    # At position 1, should see rotation
    x_rot_1 = apply_rope_i8(x, 1)
    print(f"RoPE at pos 1: {x[:4]} -> {x_rot_1[:4]}")

    # Test attention
    q = np.random.randint(-64, 64, HEAD_DIM, dtype=np.int8)
    k = np.random.randint(-64, 64, HEAD_DIM, dtype=np.int8)
    v = np.random.randint(-64, 64, HEAD_DIM, dtype=np.int8)

    output = gqa_attention_mqa_i8(q, k, v, 0)
    print(f"Attention output shape: {output.shape}")
    print(f"Attention output sample: {output[:8]}")

    # Test with KV cache
    cache = KVCache.create()
    for i in range(4):
        q_i = np.random.randint(-64, 64, HEAD_DIM, dtype=np.int8)
        k_i = np.random.randint(-64, 64, HEAD_DIM, dtype=np.int8)
        v_i = np.random.randint(-64, 64, HEAD_DIM, dtype=np.int8)
        out_i = gqa_attention_mqa_i8(q_i, k_i, v_i, i, cache)
        print(f"Position {i}: cache.seq_len = {cache.seq_len}")

    print("All attention tests passed!")

File: rust/src/attention.rs
//! GQA (Grouped Query Attention) with integer RoPE and online softmax.
//!
//! Matches the CUDA implementation in attention.cu.

use crate::linear::linear_i8_to_i32;
use crate::lut::{exp_q16_from_neg_fixed_with_coef, Exp2LutQ16, RopeLut, ROPE_HALF_DIM, ROPE_HEAD_DIM};
use crate::primitives::{clamp_i8, div_rne_tte_s64_to_s32, sra_rne_tte_s32, sra_rne_tte_s64_to_s32};

/// Constants for attention
pub const HEAD_DIM: usize = 64;
pub const MAX_SEQ_LEN: usize = 64;
pub const TOKENS_PER_BATCH: usize = 16;

// Score to exp mapping coefficient for Q0.14 scores
// LOG2E = 1.4426950408889634
// ATTN_COEF = LOG2E / 64.0
// ATTN_COEF_Q24 = round(ATTN_COEF * 2^24)
pub const LOG2E: f64 = 1.4426950408889634;
pub const ATTN_COEF: f64 = LOG2E / 64.0;
pub const ATTN_COEF_Q24: u32 = ((ATTN_COEF * (1u64 << 24) as f64) + 0.5) as u32;

// Q and sqrt(d) shifts
pub const Q_SHIFT: u32 = 2;
pub const SQRT_HEAD_DIM_SHIFT: u32 = 3;

/// Apply RoPE rotation to a vector in-place.
///
/// Rotates pairs of dimensions using precomputed cos/sin values.
/// Input/output are in int8 Q0.7 format.
pub fn apply_rope_i8(x: &mut [i8], rope_lut: &RopeLut, pos: usize) {
    assert_eq!(x.len(), HEAD_DIM);

    for i in 0..ROPE_HALF_DIM {
        let xv = x[i] as i32;
        let yv = x[i + ROPE_HALF_DIM] as i32;

        let c = rope_lut.cos(pos, i) as i32;
        let s = rope_lut.sin(pos, i) as i32;

        // Rotate: [x', y'] = [x*c - y*s, x*s + y*c]
        let x_rot_raw = xv * c - yv * s;
        let y_rot_raw = xv * s + yv * c;

        // Shift right by 15 with ties-to-even (Q15 -> Q0)
        let x_rot = sra_rne_tte_s32(x_rot_raw, 15);
        let y_rot = sra_rne_tte_s32(y_rot_raw, 15);

        x[i] = clamp_i8(x_rot);
        x[i + ROPE_HALF_DIM] = clamp_i8(y_rot);
    }
}

/// Apply RoPE to Q with additional scaling for attention.
pub fn apply_rope_q_i8(x: &mut [i8], rope_lut: &RopeLut, pos: usize) {
    assert_eq!(x.len(), HEAD_DIM);

    const NET_SHIFT: i32 = (SQRT_HEAD_DIM_SHIFT as i32) - (Q_SHIFT as i32);

    for i in 0..ROPE_HALF_DIM {
        let xv = x[i] as i32;
        let yv = x[i + ROPE_HALF_DIM] as i32;

        let c = rope_lut.cos(pos, i) as i32;
        let s = rope_lut.sin(pos, i) as i32;

        let x_rot_raw = xv * c - yv * s;
        let y_rot_raw = xv * s + yv * c;

        let mut x_rot = sra_rne_tte_s32(x_rot_raw, 15);
        let mut y_rot = sra_rne_tte_s32(y_rot_raw, 15);

        // Apply net shift
        if NET_SHIFT > 0 {
            x_rot = sra_rne_tte_s32(x_rot, NET_SHIFT as u32);
            y_rot = sra_rne_tte_s32(y_rot, NET_SHIFT as u32);
        } else if NET_SHIFT < 0 {
            x_rot <<= (-NET_SHIFT) as u32;
            y_rot <<= (-NET_SHIFT) as u32;
        }

        x[i] = clamp_i8(x_rot);
        x[i + ROPE_HALF_DIM] = clamp_i8(y_rot);
    }
}

/// Compute Q*K dot product for int8 vectors.
///
/// Returns the raw dot product (no scaling yet).
pub fn qk_dot_i8(q: &[i8], k: &[i8]) -> i32 {
    assert_eq!(q.len(), HEAD_DIM);
    assert_eq!(k.len(), HEAD_DIM);

    let mut dot: i32 = 0;
    for i in 0..HEAD_DIM {
        dot += (q[i] as i32) * (k[i] as i32);
    }
    dot
}

/// KV Cache for a single layer.
#[derive(Clone)]
pub struct KVCache {
    /// K cache: [batch, max_seq_len, head_dim]
    pub k: Vec<i8>,
    /// V cache: [batch, max_seq_len, head_dim]
    pub v: Vec<i8>,
    pub batch_size: usize,
    pub max_seq_len: usize,
    pub head_dim: usize,
}

impl KVCache {
    pub fn new(batch_size: usize, max_seq_len: usize, head_dim: usize) -> Self {
        let size = batch_size * max_seq_len * head_dim;
        Self {
            k: vec![0i8; size],
            v: vec![0i8; size],
            batch_size,
            max_seq_len,
            head_dim,
        }
    }

    /// Get K for a specific batch and position.
    pub fn get_k(&self, batch: usize, pos: usize) -> &[i8] {
        let start = (batch * self.max_seq_len + pos) * self.head_dim;
        &self.k[start..start + self.head_dim]
    }

    /// Get V for a specific batch and position.
    pub fn get_v(&self, batch: usize, pos: usize) -> &[i8] {
        let start = (batch * self.max_seq_len + pos) * self.head_dim;
        &self.v[start..start + self.head_dim]
    }

    /// Set K for a specific batch and position.
    pub fn set_k(&mut self, batch: usize, pos: usize, k: &[i8]) {
        let start = (batch * self.max_seq_len + pos) * self.head_dim;
        self.k[start..start + self.head_dim].copy_from_slice(k);
    }

    /// Set V for a specific batch and position.
    pub fn set_v(&mut self, batch: usize, pos: usize, v: &[i8]) {
        let start = (batch * self.max_seq_len + pos) * self.head_dim;
        self.v[start..start + self.head_dim].copy_from_slice(v);
    }
}

/// GQA Attention forward pass for a single step.
///
/// This implements Multi-Query Attention (MQA) where n_kv_heads = 1.
///
/// Input:
/// - qkv: [B, D + 2*kv_dim] int8 Q0.7 (output of QKV projection)
/// - kv_cache: mutable KV cache
/// - step: current position (0-indexed)
/// - score_mul_q15: [n_heads] scaling factors in Q0.15
/// - rope_lut: precomputed RoPE LUT
/// - exp_lut: precomputed exp2 LUT
///
/// Output: [B, D] int8 Q0.7
pub fn gqa_attention_mqa_i8(
    qkv: &[i8],
    kv_cache: &mut KVCache,
    step: usize,
    n_heads: usize,
    score_mul_q15: &[i32],
    rope_lut: &RopeLut,
    exp_lut: &Exp2LutQ16,
) -> Vec<i8> {
    assert!(n_heads == 4 || n_heads == 6 || n_heads == 8);
    let d = n_heads * HEAD_DIM;
    let kv_dim = HEAD_DIM; // MQA: single KV head

    let b = qkv.len() / (d + 2 * kv_dim);
    let tokens = step + 1;

    let mut out = vec![0i8; b * d];

    for batch in 0..b {
        let qkv_start = batch * (d + 2 * kv_dim);
        let k_offset = d;
        let v_offset = d + kv_dim;

        // Extract and rotate K, store to cache
        let mut k_rot = [0i8; HEAD_DIM];
        k_rot.copy_from_slice(&qkv[qkv_start + k_offset..qkv_start + k_offset + HEAD_DIM]);
        apply_rope_i8(&mut k_rot, rope_lut, step);
        kv_cache.set_k(batch, step, &k_rot);

        // Store V to cache (no rotation)
        let v_slice = &qkv[qkv_start + v_offset..qkv_start + v_offset + HEAD_DIM];
        kv_cache.set_v(batch, step, v_slice);

        // Process each head
        for h in 0..n_heads {
            // Extract and rotate Q for this head
            let q_offset = h * HEAD_DIM;
            let mut q_rot = [0i8; HEAD_DIM];
            q_rot.copy_from_slice(&qkv[qkv_start + q_offset..qkv_start + q_offset + HEAD_DIM]);
            apply_rope_q_i8(&mut q_rot, rope_lut, step);

            // Online softmax state
            let mut m: i32 = i32::MIN; // running max score
            let mut s: u64 = 0; // running sum of exp weights
            let mut vacc = [0i64; HEAD_DIM]; // weighted V accumulator

            // Compute attention over all tokens
            for t in 0..tokens {
                let k = kv_cache.get_k(batch, t);

                // Compute Q*K dot product
                let dot = qk_dot_i8(&q_rot, k);

                // Scale: dot_unshift = round(dot >> Q_SHIFT)
                let dot_unshift = sra_rne_tte_s32(dot, Q_SHIFT);

                // Apply score multiplier: score = round(dot_unshift * mul >> 15)
                let mul = score_mul_q15[h];
                let prod = (dot_unshift as i64) * (mul as i64);
                let score = sra_rne_tte_s64_to_s32(prod, 15);

                // Update online softmax
                if score > m {
                    // New max found - need to rescale
                    if m != i32::MIN {
                        let diff = m - score;
                        let scale_old = exp_q16_from_neg_fixed_with_coef(diff, exp_lut, ATTN_COEF_Q24);
                        // Rescale S
                        s = (s * scale_old as u64 + 0x8000) >> 16;
                        // Rescale vacc - must match CUDA exactly
                        // CUDA: prod += (prod >= 0) ? (1LL << 15) : -(1LL << 15); vacc = prod >> 16;
                        for v in vacc.iter_mut() {
                            let prod = (*v) * (scale_old as i64);
                            // Match CUDA's rounding: add/subtract half then arithmetic shift
                            let adjusted = if prod >= 0 {
                                prod + (1i64 << 15)
                            } else {
                                prod - (1i64 << 15)
                            };
                            *v = adjusted >> 16;
                        }
                    }
                    m = score;
                }

                // Compute exp weight for this token
                let diff = score - m;
                let w = exp_q16_from_neg_fixed_with_coef(diff, exp_lut, ATTN_COEF_Q24);
                s += w as u64;

                // Accumulate weighted V
                let v = kv_cache.get_v(batch, t);
                for i in 0..HEAD_DIM {
                    vacc[i] += (w as i64) * (v[i] as i64);
                }
            }

            // Normalize output
            let denom = if s == 0 { 1 } else { s };
            let out_start = batch * d + h * HEAD_DIM;
            for i in 0..HEAD_DIM {
                let y = div_rne_tte_s64_to_s32(vacc[i], denom);
                out[out_start + i] = clamp_i8(y);
            }
        }
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_apply_rope_identity_at_pos0() {
        let rope_lut = RopeLut::new(10000.0);
        let mut x = vec![64i8; HEAD_DIM];

        // At position 0, cos=1, sin=0, so rotation should be identity
        apply_rope_i8(&mut x, &rope_lut, 0);

        // Values should be close to original (within rounding)
        for &v in &x {
            assert!((v as i32 - 64).abs() <= 1);
        }
    }

    #[test]
    fn test_qk_dot() {
        let q = vec![10i8; HEAD_DIM];
        let k = vec![10i8; HEAD_DIM];

        let dot = qk_dot_i8(&q, &k);
        assert_eq!(dot, 10 * 10 * HEAD_DIM as i32);
    }

    #[test]
    fn test_kv_cache() {
        let mut cache = KVCache::new(2, MAX_SEQ_LEN, HEAD_DIM);

        let k = vec![42i8; HEAD_DIM];
        cache.set_k(0, 5, &k);

        let retrieved = cache.get_k(0, 5);
        assert_eq!(retrieved, k.as_slice());
    }

    #[test]
    fn test_attn_coef() {
        // Verify the coefficient is computed correctly
        let expected = ((LOG2E / 64.0) * (1u64 << 24) as f64 + 0.5) as u32;
        assert_eq!(ATTN_COEF_Q24, expected);
    }
}
